File: src/parse.py
import re
from itertools import count, groupby

from src.recursiveDescent import Parser, a, anyof, maybe, parser, skip, someof

def natural_sort(l):
	convert = lambda text: int(text) if text.isdigit() else text.lower()
	alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', str(key)) ]
	return sorted(l, key = alphanum_key)

def to_list(parse_tree):
	vals = []
	for i in parse_tree.items:
		if type(i) == parser.Node:
			vals.extend(to_list(i))
		else:
			value = i.value
			try:
				vals.append(int(value))
			except:
				try:
					ends = value.split("-")
					temp = []
					for i in range(int(ends[0].strip()), int(ends[1].strip())+1):
						temp.append(i)
					vals.extend(temp)
				except:
					pass
	return vals
			

def parse(string):
	try:
		tokens= (("[:;,]", "or"),("[^:;,&]+", "dex"), ("&", "and"))
		gram = {
			"EXPR": a(
				"VALUE", maybe(someof(skip("or"), "VALUE"))
			)
			,"VALUE": anyof("dex")
			}

		parser = Parser(tokens, gram)
		result = parser.parse("EXPR", string)
		return to_list(result)
	except:
		return None
	

def condense(vals):
	if vals is None:
		return "error"
	vals = [int(val) for val in vals]
	vals = natural_sort(list(set(vals)))
	groups = groupby(vals, key=lambda item, c=count():item-next(c))
	tmp = [list(g) for k, g in groups]
	tmp = [str(x[0]) if len(x) == 1 else "{}-{}".format(x[0],x[-1]) for x in tmp]
	return ",".join(tmp)
	
def process(string):
	tmp = parse(string)
	return condense(tmp)

def combine(stringList):
	temp = []
	for string in stringList:
		temp.extend(parse(string))
	return condense(temp)

def invert(string):
	tmp = parse(string)
	out = []
	for i in range(1,810):
		if not i in tmp:
			out.append(i)
	return condense(out)

File: src/js/parse.js
class Token {
    constructor(type, value) {
        this.type = type;
        this.value = value;
    }
}

class Lexer {
    constructor(input) {
        this.input = input;
        this.pos = 0;
    }

    nextToken() {
        while (this.pos < this.input.length && this.input[this.pos] === ' ') {
            this.pos++;
        }

        if (this.pos >= this.input.length) {
            return new Token('EOF', null);
        }

        let match = this.input[this.pos];
        this.pos++;

        // Operators
        if (match === '-' || match === '!' || match === ',' || match === '&') {
            return new Token(match, match);
        }

        // Numbers
        else if (/\d/.test(match)) {
            while (this.pos < this.input.length && /\d/.test(this.input[this.pos])) {
                match += this.input[this.pos];
                this.pos++;
            }
            return new Token('NUMBER', parseInt(match));
        }

        // Tags
        else if (/[a-zA-Z]/.test(match)) {
            while (this.pos < this.input.length && /[a-zA-Z0-9_-]/.test(this.input[this.pos])) {
                match += this.input[this.pos];
                this.pos++;
            }
            return new Token('TAG', match.toLowerCase());
        } else {
            throw new Error(`Unexpected character: ${match}`);
        }
    }
}

class Parser {
    constructor(lexer, tagMap, ids) {
        this.lexer = lexer;
        this.currentToken = this.lexer.nextToken();
        this.tagMap = tagMap;
        this.ids = ids;
    }

    eat(tokenType) {
        if (this.currentToken.type === tokenType) {
            this.currentToken = this.lexer.nextToken();
        } else {
            throw new Error(`Token of type ${tokenType} expected, got ${this.currentToken.type}`);
        }
    }

    parse() {
        return this.andExpression();
    }

    andExpression() {
        let result = this.orExpression();
        while (this.currentToken.type === '&') {
            this.eat('&');
            let term = this.orExpression();
            result = result.filter(value => term.includes(value));
        }
        return result;
    }

    orExpression() {
        let result = this.baseExpression();
        while (this.currentToken.type === ',') {
            this.eat(',');
            result = result.concat(this.baseExpression());
        }
        return result;
    }

    baseExpression() {
        if (this.currentToken.type === '!') {
            this.eat('!');
            let factor = this.baseExpression();
            let result = [];
            for (let id of this.ids) {
                if (!factor.includes(id)) {
                    result.push(id);
                }
            }
            return result;
        } else {
            return this.value();
        }
    }

    value() {
        if (this.currentToken.type === 'NUMBER') {
            let start = this.currentToken.value;
            this.eat('NUMBER');
            if (this.currentToken.type === '-') {
                this.eat('-');
                if (this.currentToken.type === 'NUMBER') {
                    let end = this.currentToken.value;
                    this.eat('NUMBER');
                    return this.ids.filter(id => {
                        const numPart = parseInt(id);
                        return numPart >= start && numPart <= end;
                    });
                } else {
                    const maxDex = Math.max(...this.ids.map(id => parseInt(id)));
                    return this.ids.filter(id => {
                        const numPart = parseInt(id);
                        return numPart >= start && numPart <= maxDex;
                    });
                }
            } else {
                //check
                return this.ids.filter(id => parseInt(id) === start);
            }
        } else if (this.currentToken.type === 'TAG') {
            const tagName = this.currentToken.value;
            this.eat('TAG');
            var numbers = this.tagMap[tagName];
            if (!numbers) {
                numbers = []
            }
            return numbers.slice();
        } else {
            throw new Error(`Unexpected token: ${this.currentToken.type}`);
        }
    }
}

function parse(input) {
    let lexer = new Lexer(input);
    let parser = new Parser(lexer, tagMap, ids);
    return parser.parse();
}

function natural_sort(l) {
    const convert = text => isNaN(text) ? text.toLowerCase() : parseInt(text);
    const alphanum_key = key => String(key).split(/(\d+)/).map(convert);
    return l.sort((a, b) => {
        const aKey = alphanum_key(a);
        const bKey = alphanum_key(b);
        for (let i = 0; i < aKey.length; i++) {
            if (aKey[i] < bKey[i]) return -1;
            if (aKey[i] > bKey[i]) return 1;
        }
        return 0;
    });
}

function condense(vals) {
    if (vals === null) {
        return "error";
    }
    vals = vals.map(val => parseInt(val));
    vals = natural_sort(Array.from(new Set(vals)));
    let groups = [];
    for (let i = 0, j = 0; i < vals.length; i++) {
        if (i === 0 || vals[i] - vals[i - 1] === 1) {
            if (!groups[j]) {
                groups[j] = [];
            }
            groups[j].push(vals[i]);
        } else {
            j++;
            groups[j] = [vals[i]];
        }
    }
    let tmp = groups.map(g => g.length === 1 ? String(g[0]) : `${g[0]}-${g[g.length - 1]}`);
    return tmp.join(",");
}
